src/v/kafka/server/handlers/fetch.cc

```diff
@@ -106,8 +131,20 @@ static ss::future<read_result> read_from_partition(
     std::unique_ptr<iobuf> data;
     std::vector<cluster::rm_stm::tx_range> aborted_transactions;
     try {
-        auto result = co_await rdr.reader.consume(
-          kafka_batch_serializer(), deadline ? *deadline : model::no_timeout);
+        const auto ts_deadline = deadline.value_or(model::no_timeout);
+        const auto failsafe_deadline = failsafe_deadline_from(ts_deadline);
+        auto consume_fut = rdr.reader.consume(
+          kafka_batch_serializer(), ts_deadline);
+        auto result
+          = co_await ss::with_timeout(failsafe_deadline, std::move(consume_fut))
```

Comment:
seastar::with_timeout usage is always worth double checking about.
1. when the timeout occurs, the future does not get cancelled. it just runs in the background. thus, it is critical that the future being wrapped in with_timeout needs to consider the lifetime of anything it captures, just like any other future that we background. in this case it sorta looks safe--I see a reference to the partition and an abort source, both of which should have lifetimes that survive this handler.
2. do we have some sort of backpressure mechanism? for example, what happens if every single time a request arrives here it times out because the readers are stuck? are those large? what happens many of them accumulate?

Reply:
About (2) I guess this will swap connection exhaustion for OOM. Maybe that's OK though? When a system gets into this state it is probably bound to fail somehow, the really important part is to get the diagnostic info.

Reply:
Thanks for the context regarding (2)